Prevent incorrect usage of ModelRenderer.

RoosterDragon opened the pull request:
Failing to Begin and End frames causes any queued rendering work to crash as buffers will not have been mapped.

pchote commented on OpenRA.Mods.Common/Traits/World/EditorActorPreview.cs:
```diff
@@ -80,16 +80,19 @@ public EditorActorPreview(WorldRenderer worldRenderer, string id, ActorReference
 				.Select(rr => rr.PrepareRender(worldRenderer));
 
-			if (r.Any())
+			Game.Renderer.WorldModelRenderer.BeginFrame();
```
This really doesn't seem right.  The ModelRenderer shouldn't be needed for the bounds calculation!

pchote replied:
Calling BeginFrame/EndFrame from anywhere but the world renderer also feels quite wrong.  We should instead try and defer `PrepareRender` until `Render` (where we should be inside a frame), and then identify what plumbing needs to change so that we can immediately calculate Bounds on the non-finalized renderables.

RoosterDragon replied:
The ModelRenderer doesn't know you're only borrowing it to calculate bounds and not actually render anything. I think it should be well within it's rights to assume you're calling RenderAsync with the intent of actually rendering stuff.

An alternative would be to silently drop any rendering requests if you're not in a frame (i.e. don't queue a render action or create the sheets required) but still give you the rendering info but that seems like a horrible gotcha.

pchote replied:
The standard `IRenderable` actually isn't actually expected to know how to render itself.  It needs to call `PrepareRender` to get an `IFinalizedRenderable` first for that.

If we defer the `PrepareRender` until we actually need it we can then use 493c9aab79d03f7f83abd8ebd2b4e5ffec28de1f (rescued out of my reflog) to query the bounds directly from the non-final renderable... hopefully.

pchote replied:
(note that I edited the comment above - we don't need to cache anything, just remove the PrepareRender)

RoosterDragon replied:
In that case, switching to "be prepared to not render when not in a frame" sounds fine - I'll switch to that.

RoosterDragon commented on OpenRA.Game/Graphics/ModelRenderer.cs:
```diff
@@ -163,64 +164,70 @@ public void SetViewportParams(Size screen, float zoom, int2 scroll)
-
-			doRender.Add(Pair.New<Sheet, Action>(sprite.Sheet, () =>
+			Sprite sprite = null;
+			Sprite shadowSprite = null;
```
If not in a frame, I've chosen to leave the sprites null here. We could still set them if required, it's just the values inside them referencing where they live on the sheet would be bogus since the sheet will be reused when the next frame starts.

reaperrr commented:
> OpenRA.Game/Graphics/ModelRenderer.cs:L186: [SA1507] The code must not contain multiple blank lines in a row.

pchote commented:
I fixed the root cause of the crash in #14564, but that doesn't really supersede this.
IMO it would now be better to throw an obvious exception if `RenderAsync` is called outside of a frame, though.

RoosterDragon committed:
Prevent incorrect usage of ModelRenderer.

Force BeginFrame/EndFrame to be called correctly, and prevent calls to RenderAsync when outside of a frame.

RoosterDragon commented:
Updated.

penev92 commented on OpenRA.Game/Graphics/ModelRenderer.cs:
```diff
@@ -279,6 +283,10 @@ static float[] ExtractRotationVector(float[] mtx)
+			if (isInFrame)
+				throw new InvalidOperationException(
+					"BeginFrame has already been called. A new frame cannot be started until EndFrame has been called.");
+			isInFrame = true;
```
Please add an empty line before this. I'm also personally not a big fan of the line break in the exception statement.

penev92 commented on OpenRA.Game/Graphics/ModelRenderer.cs:
```diff
@@ -303,6 +311,10 @@ void DisableFrameBuffer(IFrameBuffer fbo)
+			if (!isInFrame)
+				throw new InvalidOperationException(
+					"BeginFrame has not been called. There is no frame to end.");
+			isInFrame = false;
```
Same here.

RoosterDragon replied:
Done-a-roo.
